BrainPortal/lib/license_custom.rb

```diff
@@ -0,0 +1,157 @@
+  def self.included(includer) #:nodoc:
+    unless includer <= ApplicationRecord
+      raise "#{includer} is not an ActiveRecord model. The LicenseCustom module cannot be used with it."
+    end
+
+    includer.class_eval do
+      # License agreement is a pseudo attributes and cannot be accessed if the object is not saved.
+      validate   :valid_license_agreements?
+      after_save :register_license_agreements
+    end
+  end
+
+  # Returns the list of license agreements that exists for this object.
+  def license_agreements
+    @license_agreements ||= load_license_agreements
+  end
+
+  # Sets the list of license agreements that exists for this object.
+  # Can be provided with an array of license names, or a single string with
+  # a space-or-comma-separated list of license names.
+  def license_agreements=(agreements)
+    license_agreements # loads them if not already loaded
+    agrs = agreements
+    unless agrs.is_a? Array
+      agrs = agrs.to_s.split(/[,\s]+/)
+    end
+    agrs = agrs.map { |a| a.sub(/\.html\z/, "").gsub(/[^\w-]+/, "") }.uniq.sort
+    @license_agreements = agrs
+  end
+
+  # create a license file
+  def create_license_file(content, user, basename=nil)
+    basename ||= "license_#{SecureRandom.uuid().to_s}.txt"
+    userfile = create_file(content, user, basename)
+    license_agreements << userfile.id.to_s
+    register_license_agreements
+    # post save trigger
+  end
+
+  protected
+
+  # Loads the licenses from the meta data store.
+  def load_license_agreements
+    @license_agreements           = self.new_record? ? [] : self.meta[:license_agreements].presence || []
+    @_license_agreements_original = @license_agreements.dup # to determine if anything changed when saving
+    @license_agreements
+  end
+
+  # Returns true if the set of licenses are identifiers that
+  # properly match files on the filesystem, in public/licenses/{name}.html
+  def valid_license_agreements?
+    invalid_licenses = license_agreements.select do |license|
+      Userfile.find(license) 
+    end
+
+    if invalid_licenses.presence
+      invalid_licenses_list = invalid_licenses.join(", ")
+      self.errors.add(:base, "Some licence agreement files do not exist: #{invalid_licenses_list}\nPlease inform the maintainers or admins to fix issues with files.\n")
+      return false
+    end
+    return true
+  end
+
+  def create_file(content, user, basename) #:nodoc:
+    
+    file_type = TextFile
+    
+    # Temp file where the data is saved 
+    tmpcontentfile     = "/tmp/#{Process.pid}-#{rand(10000).to_s}-#{basename}" # basename's extension is used later on
+
+    File.open(tmpcontentfile, "w+") do |f|
+      f.write(content)
+    end
+
+    dest_dp_id   = DataProvider.find_by_id(user.meta["pref_data_provider_id"]).try(:id)
+    dest_dp_id ||= DataProvider.find_all_accessible_by_user(user).where(:online => true).first.try(:id)
+
+    # set to official DataProvider for Neurohub once it is ready?
```

Comment:
Not necessary it should be a general features of CBRAIN even if it was developed for Neurohub.

Reply:
perhaps a different provider for rest of users ...

Reply:
I am not sure that scope of work should include CBRAIN at this stage.